scipy/sparse/csgraph/tests/test_graph_laplacian.py

```diff
@@ -99,86 +100,244 @@ def _assert_allclose_sparse(a, b, **kwargs):
+    else:
+        assert L.dtype == dtype
+        assert d.dtype == dtype
+        desired_L = np.asarray(desired_L).astype(dtype)
+        desired_d = np.asarray(desired_d).astype(dtype)
+        _assert_allclose_sparse(L, desired_L, atol=1e-12)
+        _assert_allclose_sparse(d, desired_d, atol=1e-12)
+
+    if not copy:
+        if not (normed and check_int_type(mat)):
+            if type(mat) is np.ndarray:
+                assert_allclose(L, mat)
+            elif mat.format == "coo":
+                _assert_allclose_sparse(L, mat)
+
+
+def _check_laplacian_dtype(
+    A, desired_L, desired_d, normed, use_out_degree, copy, dtype, arr_type
+):
     mat = arr_type(A, dtype=dtype)
-    L, d = csgraph.laplacian(mat, normed=normed, return_diag=True,
-                             use_out_degree=use_out_degree,
-                             copy=copy)
+    L, d = csgraph.laplacian(
+        mat,
+        normed=normed,
+        return_diag=True,
+        use_out_degree=use_out_degree,
+        copy=copy,
+        dtype=dtype,
+    )
+    assert L.dtype == dtype
+    assert d.dtype == dtype
+    desired_L = np.asarray(desired_L).astype(dtype)
+    desired_d = np.asarray(desired_d).astype(dtype)
     _assert_allclose_sparse(L, desired_L, atol=1e-12)
     _assert_allclose_sparse(d, desired_d, atol=1e-12)
 
     if not copy:
-        if not (normed and (np.issubdtype(mat.dtype, np.signedinteger)
-                            or np.issubdtype(mat.dtype, np.uint))):
+        if not (normed and check_int_type(mat)):
             if type(mat) is np.ndarray:
                 assert_allclose(L, mat)
-            elif mat.format == 'coo':
+            elif mat.format == "coo":
                 _assert_allclose_sparse(L, mat)
 
 
-REAL_DTYPES = {np.intc, np.int_, np.longlong,
-               np.single, np.double, np.longdouble}
+INT_DTYPES = {np.intc, np.int_, np.longlong}
+REAL_DTYPES = {np.single, np.double, np.longdouble}
 COMPLEX_DTYPES = {np.csingle, np.cdouble, np.clongdouble}
 # use sorted tuple to ensure fixed order of tests
-DTYPES = tuple(sorted(REAL_DTYPES ^ COMPLEX_DTYPES, key=str))
+DTYPES = tuple(sorted(INT_DTYPES ^ REAL_DTYPES ^ COMPLEX_DTYPES, key=str))
 
 
 @pytest.mark.parametrize("dtype", DTYPES)
-@pytest.mark.parametrize("arr_type", [np.array,
-                                      sparse.csr_matrix,
-                                      sparse.coo_matrix])
+@pytest.mark.parametrize(
+    "arr_type", [np.array, sparse.csr_matrix, sparse.coo_matrix]
+)
 @pytest.mark.parametrize("copy", [True, False])
 @pytest.mark.parametrize("normed", [True, False])
 @pytest.mark.parametrize("use_out_degree", [True, False])
-def test_asymmetric_laplacian(use_out_degree, normed,
-                              copy, dtype, arr_type):
+def test_asymmetric_laplacian(use_out_degree, normed, copy, dtype, arr_type):
     # adjacency matrix
-    A = [[0, 1, 0],
-         [4, 2, 0],
-         [0, 0, 0]]
+    A = [[0, 1, 0], [4, 2, 0], [0, 0, 0]]
```

Comment:
All style changes are not necessary which maybe account for half of the diff in the PR. We tend to not reformat whole files to preserve a useful history per line (and help reviews). It's usually fine to update the style on the function you are working on, but a lot of changes in this PR are not restricted to the proposed changes. I know you are probably just using black or another formatter automatically, but we cannot do this with SciPy.

I don't want to block this PR for that, given there is a maintainer approval, but please consider this next time.

Reply:
To be clear, we don't have a specific "tradition" 😅 But because we decided not to use Black for now, we cannot have people refactor whole files with Black as we go because it breaks history. New code can follow Black's formatting but only because compliant with pep8 and our linters.

If we merge your PR right now, and someone wants to know why one particular line is written like that, they will see your PR first. Then they will have to go trough the diff and lose some time to understand that the information they are looking for is not there and go to the next commit. SciPy is 20 years old, and styling changed a few time in this timeframe. Imagine if we reformat things with all style updates the community went through, we would be burying the real nature of the change deeper and deeper as we go.

With Black and sklearn, they made the decision at the package level and enforce it. Meaning they can mark specific commits to be style changes/Black only and hence not have this issue.

Reply:
I understand. 

I have black autoformat in my VS code on every save... I'll try to turn in off.

Reply:
Either this or only stage the lines you want to push to SciPy. Since you are using VSCode, you can manually select which lines to stage.

Reply:
@mdhaber could you please help with this? I have no taste for good/bad formatting and also would have difficulty separate necessary vs unnecessary formatting changes. To my recollection, some original text would not pass lint so formating was unavoidable. 

Everything else seems to work fine: all tests pass, and I have updated the docstring format (my other weakness) as proposed by @tupui .

Reply:
@mdhaber if you are up for, updating would be better. I could have a look tomorrow too. With a visual git it's fairly easy, I just go over the changes and reverts the lies in chunks with one click.

Reply:
Sure no problem 😉

Reply:
Done. The diff is more reasonable to review.